File: sdk/mobile/flutter/lib/src/captcha_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'captcha_result.dart';
import 'captcha_config.dart';

class CaptchaClient {
  final String baseUrl;
  final String appId;
  final String appSecret;
  final CaptchaConfig config;
  final http.Client _httpClient;

  CaptchaClient({
    required this.baseUrl,
    required this.appId,
    required this.appSecret,
    CaptchaConfig? config,
  })  : this.config = config ?? CaptchaConfig(),
        _httpClient = http.Client();

  Future<SliderCaptchaResult> generateSliderCaptcha({
    int? width,
    int? height,
  }) async {
    final url = Uri.parse('$baseUrl/api/captcha/slider');

    final response = await _httpClient.post(
      url,
      headers: {
        'Content-Type': 'application/json',
        'User-Agent': 'HjtpxCaptcha-Flutter/1.0',
      },
      body: jsonEncode({
        'app_id': appId,
        'captcha_type': 'slider',
        'width': width ?? config.captchaWidth,
        'height': height ?? config.captchaHeight,
      }),
    ).timeout(Duration(seconds: config.timeout));

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return SliderCaptchaResult.fromJson(data);
    } else {
      throw CaptchaException(
        'Failed to generate captcha: ${response.statusCode}',
        response.statusCode,
      );
    }
  }

  Future<VerifyResult> verifySliderCaptcha({
    required String sessionId,
    required double x,
  }) async {
    final url = Uri.parse('$baseUrl/api/captcha/verify/slider');

    final response = await _httpClient.post(
      url,
      headers: {
        'Content-Type': 'application/json',
        'User-Agent': 'HjtpxCaptcha-Flutter/1.0',
      },
      body: jsonEncode({
        'session_id': sessionId,
        'app_id': appId,
        'x': x,
      }),
    ).timeout(Duration(seconds: config.timeout));

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return VerifyResult.fromJson(data);
    } else {
      throw CaptchaException(
        'Failed to verify captcha: ${response.statusCode}',
        response.statusCode,
      );
    }
  }

  Future<ClickCaptchaResult> generateClickCaptcha({
    int count = 4,
  }) async {
    final url = Uri.parse('$baseUrl/api/captcha/click');

    final response = await _httpClient.post(
      url,
      headers: {
        'Content-Type': 'application/json',
        'User-Agent': 'HjtpxCaptcha-Flutter/1.0',
      },
      body: jsonEncode({
        'app_id': appId,
        'captcha_type': 'click',
        'count': count,
      }),
    ).timeout(Duration(seconds: config.timeout));

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return ClickCaptchaResult.fromJson(data);
    } else {
      throw CaptchaException(
        'Failed to generate click captcha: ${response.statusCode}',
        response.statusCode,
      );
    }
  }

  Future<VerifyResult> verifyClickCaptcha({
    required String sessionId,
    required List<int> xCoords,
    required List<int> yCoords,
  }) async {
    final url = Uri.parse('$baseUrl/api/captcha/verify/click');

    final response = await _httpClient.post(
      url,
      headers: {
        'Content-Type': 'application/json',
        'User-Agent': 'HjtpxCaptcha-Flutter/1.0',
      },
      body: jsonEncode({
        'session_id': sessionId,
        'app_id': appId,
        'x_coords': xCoords,
        'y_coords': yCoords,
      }),
    ).timeout(Duration(seconds: config.timeout));

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return VerifyResult.fromJson(data);
    } else {
      throw CaptchaException(
        'Failed to verify click captcha: ${response.statusCode}',
        response.statusCode,
      );
    }
  }

  void dispose() {
    _httpClient.close();
  }
}

class CaptchaException implements Exception {
  final String message;
  final int statusCode;

  CaptchaException(this.message, this.statusCode);

  @override
  String toString() => 'CaptchaException: $message (status: $statusCode)';
}

File: sdk/mobile/react-native/src/CaptchaClient.ts
export interface SliderCaptchaResult {
  sessionId: string;
  backgroundImage: string;
  sliderImage: string;
}

export interface ClickCaptchaResult {
  sessionId: string;
  backgroundImage: string;
  targetCount: number;
}

export interface VerifyResult {
  success: boolean;
  score: number;
  message: string;
}

export interface CaptchaConfig {
  baseUrl: string;
  appId: string;
  appSecret: string;
  timeout?: number;
}

class CaptchaClient {
  private baseUrl: string;
  private appId: string;
  private appSecret: string;
  private timeout: number;

  constructor(config: CaptchaConfig) {
    this.baseUrl = config.baseUrl;
    this.appId = config.appId;
    this.appSecret = config.appSecret;
    this.timeout = config.timeout || 30000;
  }

  private async fetch(url: string, body: object): Promise<any> {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'User-Agent': 'HjtpxCaptcha-ReactNative/1.0',
        },
        body: JSON.stringify(body),
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      clearTimeout(timeoutId);
      throw error;
    }
  }

  async generateSliderCaptcha(
    width: number = 320,
    height: number = 200
  ): Promise<SliderCaptchaResult> {
    const response = await this.fetch(`${this.baseUrl}/api/captcha/slider`, {
      app_id: this.appId,
      captcha_type: 'slider',
      width,
      height,
    });

    return {
      sessionId: response.session_id,
      backgroundImage: this.baseUrl + response.background_image,
      sliderImage: this.baseUrl + response.slider_image,
    };
  }

  async verifySliderCaptcha(
    sessionId: string,
    x: number
  ): Promise<VerifyResult> {
    const response = await this.fetch(
      `${this.baseUrl}/api/captcha/verify/slider`,
      {
        session_id: sessionId,
        app_id: this.appId,
        x,
      }
    );

    return {
      success: response.success,
      score: response.score || 0,
      message: response.message || '',
    };
  }

  async generateClickCaptcha(count: number = 4): Promise<ClickCaptchaResult> {
    const response = await this.fetch(`${this.baseUrl}/api/captcha/click`, {
      app_id: this.appId,
      captcha_type: 'click',
      count,
    });

    return {
      sessionId: response.session_id,
      backgroundImage: this.baseUrl + response.background_image,
      targetCount: response.target_count,
    };
  }

  async verifyClickCaptcha(
    sessionId: string,
    xCoords: number[],
    yCoords: number[]
  ): Promise<VerifyResult> {
    const response = await this.fetch(
      `${this.baseUrl}/api/captcha/verify/click`,
      {
        session_id: sessionId,
        app_id: this.appId,
        x_coords: xCoords,
        y_coords: yCoords,
      }
    );

    return {
      success: response.success,
      score: response.score || 0,
      message: response.message || '',
    };
  }
}

export default CaptchaClient;
